library/network/src/modules/SuSEFirewall.rb

```diff
@@ -1151,6 +1151,92 @@ def GetInterfacesInZoneSupportingAnyFeature(zone)
       GetInterfacesInZone(zone)
     end
 
+    # Function sets status for several services in several firewall zones.
+    #
+    # @param	list <string> service ids
+    # @param	list <string> firewall zones (EXT|INT|DMZ...)
+    # @param	boolean new status of services
+    # @return	[Boolean] if successfull
+    #
+    # @example
+    #	SetServicesForZones (["samba-server", "service:irc-server"], ["DMZ", "EXT"], false);
+    #	SetServicesForZones (["samba-server", "service:irc-server"], ["EXT", "DMZ"], true);
+    #
+    # @see #GetServicesInZones()
+    # @see #GetServices()
+    def SetServicesForZones(services_ids, firewall_zones, new_status)
+
+      Yast.import "SuSEFirewallServices"
+
+      services_ids = deep_copy(services_ids)
+      zones = deep_copy(firewall_zones)
+      # no groups == all groups
+      if zones.empty?
+        zones = GetKnownFirewallZones()
+      end
+
+      # Rebuild services since we may have more than one FirewallD
+      # service per SF2 one
+      if self.class == SuSEFirewalld
+        tmp_services_ids = deep_copy(services_ids)
+        services_ids = []
+        tmp_services_ids.each do |service|
+          _sf2_to_firewalld_service(service).each do |s|
+            services_ids << s
+          end
+        end
+      end
+
+      # setting for each service
+      services_ids.each do |service|
+        # Service is not supported by firewalld.
+        # We can only do such error checking if backend is running
+        if IsStarted() and not @fwd_api.get_supported_services.include?(service)
+          Builtins.y2error("Undefined service '#{service}'")
+          raise(SuSEFirewalServiceNotFound,
+                "Service with name '#{service}' does not exist")
+        end
+        zones.each do |zone|
+          # Add/remove service to/from zone only if zone is not 'trusted'.
+          # For 'trusted' zone there is no need to explicitly add/remove
+          # services as all connections are by default accepted.
+          next if zone == "trusted"
+
+          # zone must be known one
+          if !IsKnownZone(zone)
+            Builtins.y2error(
+              "Zone '%1' is unknown firewall zone, skipping...",
+              zone
+            )
+            next
+          end
+
+          if new_status == true # enable
+            Builtins.y2milestone(
+              "Adding '%1' into '%2' zone",
+              service, zone
+            )
+            # Only add it if it is not there
+            if !@SETTINGS[zone][:services].include?(service)
+              @SETTINGS[zone][:services] << service
+              SetModified()
+              @SETTINGS[zone][:modified] << :services
+            end
+          else # disable
+            Builtins.y2milestone(
+              "Removing '%1' from '%2' zone",
+               service, zone
+            )
+            @SETTINGS[zone][:services].delete(service)
+            SetModified()
+            @SETTINGS[zone][:modified] << :services
```

Comment:
All the code touching SETTINGS is begging for wrappers to encapsulate the details.